### PR TITLE
Implement inline stack probes for AArch64

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR implements necessary support for inline stack probes in the AArch64 backend and then enables the support in Wasmtime. This builds on https://github.com/bytecodealliance/wasmtime/pull/5350 to ensure that on these hosts an accidentally misconfigured `Config` will still guarantee that wasm itself will run into the guard page for the stack rather than skipping over it.

The commits here have been split out with various refactorings for inline stack probes, including a minor optimization for amodes in the aarch64 backend. I'd appreciate some close review on the aarch64 bits since I'm mostly just going with what works first and I don't necessarily fully understand the implications of the changes.

### CODE REVIEW

File: cranelift/codegen/src/isa/aarch64/abi.rs [hunk 1]
```diff
@@ -637,12 +637,46 @@ impl ABIMachineSpec for AArch64MachineDeps {
...
-        unimplemented!("Inline stack probing is unimplemented on AArch64");
+    fn gen_inline_probestack(insts: &mut SmallInstVec<Self::I>, frame_size: u32, guard_size: u32) {
+        // The stack probe loop currently takes 6 instructions and each inline
+        // probe takes 2 (ish, these numbers sort of depend on the constants).
+        // Set this to 3 to keep the max size of the probe to 6 instructions.
+        const PROBE_MAX_UNROLL: u32 = 3;
+
+        let probe_count = align_to(frame_size, guard_size) / guard_size;
+        if probe_count <= PROBE_MAX_UNROLL {
+            // When manually unrolling stick an instruciton that stores 0 at a
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/aarch64/abi.rs:648]: *instruction

**Comment by afonso360** [comment]
Can you update [fuzzgen/src/lib.rs:269](https://github.com/bytecodealliance/wasmtime/blob/8bc75502112d05e2f16e0714e59ae3cb4bd9c43c/cranelift/fuzzgen/src/lib.rs#L269-L280) and enable inline stack probing for AArch64 in the cranelift fuzzer?

File: cranelift/codegen/src/isa/aarch64/inst.isle [hunk 2]
```diff
@@ -935,7 +935,13 @@
...
        ;; A dummy use, useful to keep a value alive.
        (DummyUse
-        (reg Reg))))
+        (reg Reg))
+
+       ;; Emits a inline stack probe loop.
+       (StackProbeLoop (start WritableReg)
+                       (end Reg)
+                       (step Imm12))
+))
```
Comment (cfallin) [inline cranelift/codegen/src/isa/aarch64/inst.isle:944]: Can we keep the closing parens all together (Lisp-style)?

File: cranelift/codegen/src/isa/aarch64/inst.isle [hunk 3]
```diff
@@ -935,7 +935,13 @@
 
        ;; A dummy use, useful to keep a value alive.
        (DummyUse
-        (reg Reg))))
+        (reg Reg))
+
+       ;; Emits a inline stack probe loop.
```
Comment (cfallin) [inline cranelift/codegen/src/isa/aarch64/inst.isle:940]: s/a/an/

Also can we note here that it's careful not to use spilltmp or tmp2, so these registers are allowed to be inputs?

File: cranelift/codegen/src/isa/aarch64/inst/emit.rs [hunk 4]
```diff
@@ -3424,6 +3417,66 @@ impl MachInstEmit for Inst {
...
+
+            &Inst::StackProbeLoop { start, end, step } => {
+                assert!(emit_info.0.enable_probestack());
+                let start = allocs.next_writable(start);
+                let end = allocs.next(end);
+
+                // The loop generated here uses `start` as a counter register to
+                // count backwards until negating it exceeds `end`. In other
+                // words `start` is an offset from `sp` we're testing where
+                // `end` is the max size we need to test. The loop looks lik:
```
Comment (cfallin) [inline cranelift/codegen/src/isa/aarch64/inst/emit.rs:3429]: s/lik/like/

**Review by cfallin — APPROVED** [review]
Looks correct to me -- thanks for the careful work here! Most of the subtlety is around the invariants for use of x16/x17 (spilltmp/tmp2) as you've seen so my comments are mostly requests for more notes in the code around this.

**Comment by alexcrichton** [comment]
@cfallin mind double checking the [next-to-last commit](https://github.com/bytecodealliance/wasmtime/pull/5353/commits/44cad07a29fa4c781771fa0d702a572c7c39ab31) here now?

**Comment by kpreisser** [comment]
Nice, thank you! The repro `.wat` from #5340 now also successfully runs on Windows Arm64 (`aarch64-pc-windows-msvc`).

**Pull request merged**